File: Wallet.js
const { v4: uuidv4 } = require('uuid');
const { TransactionHistory } = require('./TransactionHistory');
const { MoneyUtils } = require('./MoneyUtils');

class Wallet {
    static MAX_TRANSACTION_AMOUNT = 200000000;

    constructor() {
        this.id = uuidv4();
        this.balance = 0.0;
        this.transactionHistory = new TransactionHistory();
    }

    deposit(amount) {
        if (MoneyUtils.isAmountValid(amount)) {
            this.balance += amount;
            this.transactionHistory.addTransaction(amount, 'DEPOSIT', 'me');
        }
    }

    withdraw(amount) {
        if (MoneyUtils.isAmountValid(amount, this.balance, Wallet.MAX_TRANSACTION_AMOUNT)) {
            this.balance -= amount;
            this.transactionHistory.addTransaction(amount, 'WITHDRAW', '');
            return true;
        }
        return false;
    }

    transfer(amount, destinationUser) {
        if (MoneyUtils.isAmountValid(amount, this.balance, Wallet.MAX_TRANSACTION_AMOUNT)) {
            this.balance -= amount;
            destinationUser.wallet.deposit(amount);
            this.transactionHistory.addTransaction(amount, 'TRANSFER', destinationUser.username);
            return true;
        }
        return false;
    }
}

module.exports = { Wallet };

File: Wallet.py
import uuid
from TransactionHistory import TransactionHistory
from MoneyUtils import MoneyUtils


class Wallet:
    MAX_TRANSACTION_AMOUNT = 200000000

    def __init__(self):
        self.id = str(uuid.uuid4())
        self.balance = 0.0
        self.transaction_history = TransactionHistory()

    def deposit(self, amount):
        if MoneyUtils.is_amount_valid(amount):
            self.balance += amount
            self.transaction_history.add_transaction(amount, "Deposit", "me")

    def withdraw(self, amount):
        if MoneyUtils.is_amount_valid(amount, self.balance, self.MAX_TRANSACTION_AMOUNT):
            self.balance -= amount
            self.transaction_history.add_transaction(amount, "Withdraw", "")
            return True
        return False

    def transfer(self, amount, destination_user):
        if MoneyUtils.is_amount_valid(amount, self.balance, self.MAX_TRANSACTION_AMOUNT):
            self.balance -= amount
            destination_user.wallet.deposit(amount)
            self.transaction_history.add_transaction(amount, "Transfer", destination_user.username)
            return True
        return False

